### PR TITLE
Improved HttpClientConfig

### PR DESCRIPTION
**Pull request by mirvnillith**
- timeoutMs
- followRedirect
- path of URL used as API root

### CODE REVIEW

File: client/src/main/java/se/fortnox/reactivewizard/client/HttpClient.java [hunk 1]
```diff
@@ -131,6 +132,7 @@ public HttpClient(HttpClientConfig config,
         collector = new ByteBufCollector(config.getMaxResponseSize());
         this.preRequestHooks = preRequestHooks;
         this.retryDuration = Duration.ofMillis(config.getRetryDelayMs());
+        setTimeout(config.getTimeoutMs(), ChronoUnit.MICROS);
```
Comment (erljef) [inline client/src/main/java/se/fortnox/reactivewizard/client/HttpClient.java:135]: Shouldn't this be `ChronoUnit.MILLIS`?

Comment (mirvnillith) [reply client/src/main/java/se/fortnox/reactivewizard/client/HttpClient.java]: Effin' code-complete (MI -> MILLIS, MICROS or MILLENIA ;)

File: client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java [hunk 2]
```diff
@@ -83,6 +86,14 @@ void shouldSetPort() throws URISyntaxException {
         assertThat(config.getHost()).isEqualTo("localhost");
     }
 
+    @Test
+    void shouldSetRoot() throws URISyntaxException {
+		assertThat(new HttpClientConfig("http://localhost:8080").getRoot()).isEqualTo("");
```
Comment (softqwewasd) [inline client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java:91]: Missing indentation here.

**Comment by softqwewasd** [comment]
Nice changes! :star2: We should probably create a new PR to deprecate or possibly remove the readTimeoutMs altogether and just use the new timeout config in the flux timeout operator and netty's timeout handler :)

File: client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java (hunk 2 above)
Comment (mirvnillith) [reply client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java:91]: How so, it looks identical to indentations above and below?

**Comment by mirvnillith** [comment]
How do I see these 2 Checkstyle errors (they also don't show when building locally)?

[INFO] --- maven-checkstyle-plugin:3.3.1:check (validate) @ reactivewizard-client ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Skipping Reactive Wizard
[INFO] This project has been banned from the build due to previous failures.
[INFO] ------------------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Reactive Wizard 999.9.9-SNAPSHOT:
[INFO] 
[INFO] Reactive Wizard .................................... SUCCESS [  1.263 s]
[INFO] reactivewizard-api ................................. SUCCESS [  2.761 s]
[INFO] reactivewizard-test ................................ SUCCESS [ 20.658 s]
[INFO] reactivewizard-utils ............................... SUCCESS [  9.661 s]
[INFO] reactivewizard-binding ............................. SUCCESS [  6.544 s]
[INFO] reactivewizard-config .............................. SUCCESS [  8.818 s]
[INFO] reactivewizard-json ................................ SUCCESS [  6.007 s]
[INFO] reactivewizard-jaxrs-api ........................... SUCCESS [  5.196 s]
[INFO] reactivewizard-jaxrs-api-test ...................... SUCCESS [  3.930 s]
[INFO] reactivewizard-metrics ............................. SUCCESS [  4.696 s]
[INFO] reactivewizard-dao ................................. SUCCESS [ 26.772 s]
[INFO] reactivewizard-jaxrs ............................... SUCCESS [ 21.623 s]
[INFO] reactivewizard-bootstrap ........................... SUCCESS [  1.306 s]
[INFO] reactivewizard-server .............................. SUCCESS [ 41.378 s]
[INFO] reactivewizard-examples ............................ SUCCESS [  0.034 s]
[INFO] hello-world ........................................ SUCCESS [  7.549 s]
[INFO] reactivewizard-validation .......................... SUCCESS [ 11.470 s]
[INFO] reactivewizard-client .............................. FAILURE [01:39 min]
[INFO] reactivewizard-dbmigrate ........................... SKIPPED
[INFO] coverage ........................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:39 min
[INFO] Finished at: 2024-05-28T12:53:04Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.1:check (validate) on project reactivewizard-client: Failed during checkstyle execution: There are 2 errors reported by Checkstyle 10.14.0 with checkstyle.xml ruleset. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :reactivewizard-client
Error: Process completed with exit code 1.

File: client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java (hunk 2 above)
Comment (softqwewasd) [reply client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java:91]: image
compared to ex.
image

Comment (mirvnillith) [reply client/src/test/java/se/fortnox/reactivewizard/client/HttpClientConfigTest.java:91]: Hmm, this is what I see:

bild

**Comment by mirvnillith** [comment]
Found and fixed the Checkstyle errors by working through every single change. It was the missing spaces around - when stripping the leading root / ...

**Review by softqwewasd — APPROVED** [review]

**Review by magnusp — APPROVED** [review]

**Review by jepp3 — APPROVED** [review]

**Comment by splitfeed** [comment]
Timeout and following redirects is pretty straight forward, but I would love a description of what the URL path settings is meant to be used for :)

**Comment by mirvnillith** [comment]
> Timeout and following redirects is pretty straight forward, but I would love a description of what the URL path settings is meant to be used for :)

When implementing an API at another root. E.g. the JAX-RS @Path is "/my/api" but the implementation is exposed as "/service/apis/my/api". Currently HttpClient is only able to call endpoints that have their full path in JAX-RS.

**Pull request merged**